spotlight/factorization/implicit.py

```diff
@@ -220,11 +219,114 @@ def fit(self, interactions, verbose=False):
 
             epoch_loss = 0.0
 
-            for (minibatch_num,
-                 (batch_user,
-                  batch_item)) in enumerate(minibatch(user_ids_tensor,
-                                                      item_ids_tensor,
-                                                      batch_size=self._batch_size)):
+            for (
+                    minibatch_num,
+                    (
+                        batch_user,
+                        batch_item,
+                    )
+            ) in enumerate(
+                minibatch(
+                    user_ids_tensor,
+                    item_ids_tensor,
+                    batch_size=self._batch_size
+                )
+            ):
+
+                positive_prediction = self._net(batch_user, batch_item)
+
+                if self._loss == 'adaptive_hinge':
+                    negative_prediction = self._get_multiple_negative_predictions(
+                        batch_user, n=self._num_negative_samples)
+                else:
+                    negative_prediction = self._get_negative_prediction(batch_user)
+
+                self._optimizer.zero_grad()
+
+                loss = self._loss_func(
+                    positive_prediction,
+                    negative_prediction,
+                )
+                epoch_loss += loss.item()
+
+                loss.backward()
+                self._optimizer.step()
+
+            epoch_loss /= minibatch_num + 1
+
+            if verbose:
+                print('Epoch {}: loss {}'.format(epoch_num, epoch_loss))
+
+            if np.isnan(epoch_loss) or epoch_loss == 0.0:
+                raise ValueError('Degenerate epoch loss: {}'
+                                 .format(epoch_loss))
+
+    def fit_weighted(self, interactions, verbose=False):
```

Comment:
I think we should keep a single `fit` method. If weights are present, we use them; if not, we don't.

This could probably be accomplished by changing the `minibatch` function to just infinitely yield `None` for those arguments that are `None` rather than tensors. This way, it always yields `None` for `batch_sample_weights` which ties in nicely with how the loss functions are changed.

Reply:
Definitely tried to keep it to a single fit function at first. But I want to avoid consuming ram for an unnecessary tensor of Nones.
Would this lead to a tensor of Nones (not None) being passed to the loss functions, rather than a single None? (unless the tensor is not created by yielding Nones?)
What if we just checked if interactions.weights is specified at the beginning of fit, and if they are just have fit call _fit_weighted?

Sequence weight question for you: I was thinking about the sequence interactions yesterday, and my guess is to make a parallel sequence tensor full of sample weights, of the same dimensions as the sequence tensor of item-ids. Does this sound right?

Reply:
No, we won't need a tensor of nones. I'll add a comment with a prototype of what I meant.